File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.py
import sys

input = sys.stdin.readline

N = int(input())
A = []
for _ in range(N):
    A.append(list(map(int,input().split())))

answer = sys.maxsize

def dfs(first, now, cnt, temp):
    global answer
    if cnt == N :
        if A[now][first] != 0:
            answer = min(answer, temp + A[now][first])

    for i in range(N):
        if now == i:
            continue
        if not visited[i] and A[now][i] != 0:
            visited[i] = True
            dfs(first, i, cnt + 1, temp + A[now][i])
            visited[i] = False

for i in range(N):
    visited = [False] * (N)
    visited[i] = True
    dfs(i,i,1,0)
    visited[i] = False

print(answer)

File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.java
import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;


public class Main {

	static long[][] grid;
	static int N;
	static boolean[] visited;
	static long answer;
	public static void main(String[] args) throws  IOException {
		// 외판원 순회2
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		N = Integer.parseInt(br.readLine());
		
		grid = new long[N][N];
		visited = new boolean[N];
		
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < N; j++) {
				grid[i][j] = Long.parseLong(st.nextToken());
			}
		}
		
		answer = Long.MAX_VALUE;
		
		for (int i = 0; i < N; i++) {
			visited[i] = true;
			dfs(i, i, 0, 0);
			visited[i] = false;
		}
		
		System.out.println(answer);
	}
	
	static void dfs(int init, int start, long total, int cnt) {
		if (cnt == N - 1) {
			if (grid[start][init] == 0) return;
			answer = Math.min(answer, total + grid[start][init]);
			return;
		}
		
		for (int i = 0; i < N; i++) {
			if (!visited[i] && grid[start][i] != 0) {
				visited[i] = true;
				dfs(init, i, total + grid[start][i], cnt +1);
				visited[i] = false;
			}
		}
	}

}
